File: server/src/main/java/dev/c0redev/volter/Poly.java
package dev.c0redev.volter;

import java.nio.charset.StandardCharsets;
import org.bouncycastle.crypto.engines.ChaCha7539Engine;
import org.bouncycastle.crypto.params.KeyParameter;
import org.bouncycastle.crypto.params.ParametersWithIV;

final class Poly {

  private final ChaCha7539Engine engine = new ChaCha7539Engine();
  private final byte[] zero8 = new byte[8];

  Poly(byte[] secret, long slot, String info) {
    byte[] key = Dexote.hkdf(secret, Dexote.slotBytes(slot), "dexote-poly-" + info, 32);
    byte[] nonce = new byte[12];
    engine.init(true, new ParametersWithIV(new KeyParameter(key), nonce));
  }

  long next64() {
    byte[] out = new byte[8];
    engine.processBytes(zero8, 0, 8, out, 0);
    long v = 0;
    for (int i = 0; i < 8; i++) v = (v << 8) | (out[i] & 0xff);
    return v;
  }

  int intRange(int min, int max) {
    if (max <= min) return min;
    long span = (long) (max - min) + 1;
    return min + (int) (Long.remainderUnsigned(next64(), span));
  }

  int padLen(int maxPad) {
    if (maxPad <= 0) return 0;
    return intRange(0, maxPad);
  }

  private static byte[] info(String s) {
    return s.getBytes(StandardCharsets.UTF_8);
  }
}

File: internal/dexote/poly.go
package dexote

import (
	"encoding/binary"

	"golang.org/x/crypto/chacha20"
)

type Poly struct {
	cipher *chacha20.Cipher
	zero   []byte
}

func NewPoly(secret []byte, slot int64, info string) *Poly {
	key := hkdfKey(secret, slotBytes(slot), "dexote-poly-"+info, 32)
	nonce := make([]byte, chacha20.NonceSize)
	c, err := chacha20.NewUnauthenticatedCipher(key, nonce)
	if err != nil {
		panic("dexote: poly cipher: " + err.Error())
	}
	return &Poly{cipher: c, zero: make([]byte, 8)}
}

func (p *Poly) next64() uint64 {
	buf := make([]byte, 8)
	p.cipher.XORKeyStream(buf, p.zero)
	return binary.BigEndian.Uint64(buf)
}

func (p *Poly) IntRange(min, max int) int {
	if max <= min {
		return min
	}
	span := uint64(max-min) + 1
	return min + int(p.next64()%span)
}

func (p *Poly) PadLen(maxPad int) int {
	if maxPad <= 0 {
		return 0
	}
	return p.IntRange(0, maxPad)
}

func (p *Poly) JunkPlan(maxCount, minSize, maxSize int) []int {
	n := p.IntRange(0, maxCount)
	sizes := make([]int, n)
	for i := range sizes {
		sizes[i] = p.IntRange(minSize, maxSize)
	}
	return sizes
}
